sdk/search/Azure.Search.Documents/src/Batching/SearchIndexingBufferedSender{T}.cs

```diff
@@ -0,0 +1,606 @@
+        internal async Task DisposeAsync(bool async)
+        {
+            if (Interlocked.CompareExchange(ref _disposed, 1, 0) == 0)
+            {
+                await _publisher.DisposeAsync(async).ConfigureAwait(false);
+                _keyFieldBarrier.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Ensure the sender was properly disposed.
+        /// </summary>
+        ~SearchIndexingBufferedSender()
+        {
+            if (_publisher?.IndexingActionsCount > 0)
+            {
+                try
+                {
+                    #pragma warning disable CA1065 // Do not raise exceptions in unexpected locations
+                    throw new InvalidOperationException(
+                        $"{nameof(SearchIndexingBufferedSender<T>)} has {_publisher.IndexingActionsCount} unsent indexing actions.");
+                    #pragma warning restore CA1065 // Do not raise exceptions in unexpected locations
+                }
+                catch (InvalidOperationException)
+                {
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensure nobody's trying to use this after it's been disposed.
+        /// </summary>
+        private void EnsureNotDisposed()
+        {
+            if (_disposed > 0)
+            {
+                throw new ObjectDisposedException(nameof(SearchIndexingBufferedSender<T>));
+            }
+        }
+        #endregion Dispose
+
+        #region Get Key Field
+        /// <summary>
+        /// Ensure we have a valid KeyFieldAccessor (and probe the type or
+        /// service to find one).
+        /// </summary>
+        /// <param name="async">Whether to run sync or async.</param>
+        /// <param name="cancellationToken">The Cancellation token.</param>
+        /// <returns>A task that will complete after we've checked.</returns>
+        private async Task EnsureKeyFieldAccessorAsync(bool async, CancellationToken cancellationToken)
+        {
+            // Skip initialization if we already have one
+            if (KeyFieldAccessor != null) { return; }
+
+            // Make sure we only run the initialization code once
+            if (await _keyFieldBarrier.TryEnterAsync(async, cancellationToken).ConfigureAwait(false))
+            {
+                // Release no matter what
+                try
+                {
+                    // Case 1: The user provided an explicit accessor and we're done
+                    if (KeyFieldAccessor != null) { return; }
+
+                    // Case 2: Infer the accessor from FieldBuilder
+                    try
+                    {
+                        FieldBuilder builder = new FieldBuilder { Serializer = SearchClient.Serializer };
+                        IDictionary<string, SearchField> fields = builder.BuildMapping(typeof(T));
+                        KeyValuePair<string, SearchField> keyField = fields.First(pair => pair.Value.IsKey == true);
```

Comment:
`FirstOrDefault` and avoid throwing in case none was defined, since you can check for "null" (does `FirstOrDefault` return nullable?).

Reply:
That's fair

Reply:
Fixed